lexicon/providers/aliyun.py

```
@@ -0,0 +1,217 @@
+    def _list_records(self, rtype=None, name=None, content=None):
+        query_params = {}
+
+        if rtype:
+            query_params['TypeKeyWord'] = rtype
+
+        if name:
+            query_params['RRKeyWord'] = self._relative_name(name)
+
+        if content:
+            query_params['ValueKeyWord'] = content
+
+        response = self._request('DescribeDomainRecords', query_params=query_params)
+
+        resource_list = response['DomainRecords']['Record']
+
+        processed_records = []
+        for resource in resource_list:
+            processed_records.append({
+                'id': resource['RecordId'],
+                'type': resource['Type'],
+                'name': self._full_name(resource['RR']),
+                'ttl': resource['TTL'],
+                'content': resource['Value']
+            })
+        LOGGER.debug('list_records: %s', processed_records)
+        return processed_records
+
+    def __get_record(self, identifier):
+        response = self._request('DescribeDomainRecordInfo', query_params={'RecordId': identifier})
+
+        return {
+            'id': response['RecordId'],
+            'type': response['Type'],
+            'name': self._full_name(response['RR']),
+            'ttl': response['TTL'],
+            'content': response['Value']
+        }
+
+    # Create or update a record.
+    def _update_record(self, identifier, rtype=None, name=None, content=None):
+        resources = self._list_records(rtype, name, None)
+
+        for record in resources:
+            if (rtype == record['type']) \
+                and (self._relative_name(name) == self._relative_name(record['name'])) \
+                and (content == record['content']):
+                return True
+
+        if not identifier:
+            record = resources[0] if resources else None
+            identifier = record['id'] if record else None
+
+        if not identifier:
+            self._create_record(rtype, name, content)
```

The contract of `update_record`, as specified in https://github.com/AnalogJ/lexicon/blob/master/SPECIFICATION.md#update_record, is to update an existing record, not to create it if it does not exist. You should throw an exception if no record to update could be found.